week3/api-services/auth_security_system.py

```
@@ -0,0 +1,682 @@
+        self.db_path = db_path
+    
+    def log_request(self, 
+                   user_id: str, 
+                   api_key: str,
+                   endpoint: str,
+                   method: str,
+                   status_code: int,
+                   response_time_ms: float,
+                   request_size: int = 0,
+                   response_size: int = 0,
+                   ip_address: str = None,
```

Ruff (0.12.2)

PEP 484 prohibits implicit `Optional`

Convert to `Optional[T]`

(RUF013)